src/plugins/profiling/server/routes/stacktrace.ts

```diff
@@ -23,36 +23,104 @@ import { getHitsItems, getDocs } from './compat';
-  const fileIDChunk = frameID.slice(0, 23);
-  let fileID = frameIDToFileIDCache.get(fileIDChunk) as string;
-  if (fileID) return fileID;
-
-  // Step 1: Convert the base64-encoded frameID to an array of 22 bytes.
-  // We use 'base64url' instead of 'base64' because frameID is encoded URL-friendly.
-  // The first 16 bytes contain the FileID.
-  const buf = Buffer.from(fileIDChunk, 'base64url');
-
-  // Convert the FileID bytes into base64 with URL-friendly encoding.
-  // We have to manually append '==' since we use the FileID string for
-  // comparing / looking up the FileID strings in the ES indices, which have
-  // the '==' appended.
-  // We may want to remove '==' in the future to reduce the uncompressed storage size by 10%.
-  fileID = buf.toString('base64url', 0, 16) + '==';
-  frameIDToFileIDCache.set(fileIDChunk, fileID);
-  return fileID;
+const FILE_ID_LENGTH = 16;
+const FRAME_ID_LENGTH = 24;
+
+export interface EncodedStackTrace {
+  // This field is a base64-encoded byte string. The string represents a
+  // serialized list of frame IDs. Each frame ID is composed of two
+  // concatenated values: a 16-byte file ID and an 8-byte address or line
+  // number (depending on the context of the downstream reader).
+  //
+  //         Frame ID #1               Frame ID #2
+  // +----------------+--------+----------------+--------+----
+  // |     File ID    |  Addr  |     File ID    |  Addr  |
+  // +----------------+--------+----------------+--------+----
+  FrameID: string;
+
+  // This field is a run-length encoding of a list of uint8s. The order is
+  // reversed from the original input.
+  Type: string;
 }
 
-// extractFileIDArrayFromFrameIDArray extracts all FileIDs from the array of FrameIDs
-// and returns them as an array of base64url encoded strings. The order of this array
-// corresponds to the order of the input array.
-function extractFileIDArrayFromFrameIDArray(frameIDs: string[]): string[] {
-  const fileIDs = Array<string>(frameIDs.length);
-  for (let i = 0; i < frameIDs.length; i++) {
-    fileIDs[i] = extractFileIDFromFrameID(frameIDs[i]);
+// runLengthDecodeReverse decodes a run-length encoding for the reversed input array.
+//
+// The input is a binary stream of 2-byte pairs (first byte is the length and the
+// second byte is the binary representation of the object). The output is a list of
+// uint8s in reverse order.
+//
+// E.g. byte array [5, 0, 2, 2] is converted into an uint8 array like
+// [2, 2, 0, 0, 0, 0, 0].
+export function runLengthDecodeReverse(input: Buffer): number[] {
+  let size = 0;
+  for (let i = 0; i < input.length; i += 2) {
+    size += input[i];
   }
-  return fileIDs;
+
+  const output: number[] = new Array(size);
+
+  let i = input.length - 2;
+  let j = 0;
+  let count = input[i];
+
+  while (j < output.length) {
+    if (count > 0) {
+      output[j] = input[i + 1];
+      count--;
+      j++;
+    } else {
+      i -= 2;
+      count = input[i];
+    }
+  }
```

Comment:
If we assume `size` to be correct, this could be simplified to
```suggestion
  let idx = 0;
  for (let i = input.length - 1; i >= 1; i -= 2) {
    for (let j = 0; j < input[i - 1]; j++) {
      output[idx] = input[i];
      idx++;
    }
  }
```
Maybe there is some JS/TS syntactic sugar to replace the inner loop with !?

Reply:
I couldn't find anything to replace the inner loop that wasn't longer than your suggestion.

I like the simplicity of your code anyway. :+1: